llm-examples/mpt/mpt-30b/02_mlflow_logging_inference.py

```diff
@@ -0,0 +1,193 @@
+            "Write a response that appropriately completes the request."
+        )
+
+        return f"""{INTRO_BLURB}
+        {INSTRUCTION_KEY}
+        {instruction}
+        {RESPONSE_KEY}
+        """
+
+    def predict(self, context, model_input):
+        """
+        This method generates prediction for the given input.
+        """
+        generated_text = []
+        for index, row in model_input.iterrows():
+          prompt = row["prompt"]
+          temperature = model_input.get("temperature", [1.0])[0]
+          max_new_tokens = model_input.get("max_new_tokens", [100])[0]
+          full_prompt = self._build_prompt(prompt)
+          encoded_input = self.tokenizer.encode(full_prompt, return_tensors="pt").to('cuda')
+          output = self.model.generate(encoded_input, do_sample=True, temperature=temperature, max_new_tokens=max_new_tokens)
+          prompt_length = len(encoded_input[0])
+          generated_text.append(self.tokenizer.batch_decode(output[:,prompt_length:], skip_special_tokens=True))
+        return pd.Series(generated_text)
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC Download the model
+
+# COMMAND ----------
+
+from huggingface_hub import snapshot_download
+
+# If the model has been downloaded in previous cells, this will not repetitively download large model files, but only the remaining files in the repo
+model_location = snapshot_download(repo_id="mosaicml/mpt-30b-instruct", cache_dir="/local_disk0/.cache/huggingface/", revision="2d1dde986c9737e0ef4fc2280ad264baf55ea1cd")
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC Log the model to MLFlow
+
+# COMMAND ----------
+
+from mlflow.models.signature import ModelSignature
+from mlflow.types import DataType, Schema, ColSpec
+
+# Define input and output schema
+input_schema = Schema([
+    ColSpec(DataType.string, "prompt"), 
+    ColSpec(DataType.double, "temperature"), 
+    ColSpec(DataType.long, "max_tokens")])
+output_schema = Schema([ColSpec(DataType.string)])
+signature = ModelSignature(inputs=input_schema, outputs=output_schema)
+
+# Define input example
+input_example=pd.DataFrame({
+            "prompt":["what is ML?"], 
+            "temperature": [0.5],
+            "max_tokens": [100]})
+
+# Log the model with its details such as artifacts, pip requirements and input example
+# This may take about 20 minutes to complete
+with mlflow.start_run() as run:  
+    mlflow.pyfunc.log_model(
+        "model",
+        python_model=MPT(),
+        artifacts={'repository' : model_location},
+        pip_requirements=[f"torch=={torch.__version__}", 
+                          f"transformers=={transformers.__version__}", 
+                          f"accelerate=={accelerate.__version__}", "einops", "sentencepiece", "xformers"],
+        input_example=input_example,
+        signature=signature
+    )
+
+# COMMAND ----------
+
+import mlflow
+import pandas as pd
+
+logged_model = 'runs:/7715695b0c134d31a30c6c5ab4c9d43f/model'
```

Comment:
Should we add a comment that this should be set by user or use `"runs:/" + run.info.run_id + "/model"`?

Reply:
The example should use the model from model registry like other examples. I will remove this. Thanks.